### PR TITLE
Add validation for Adreno GPU.

### PR DESCRIPTION
**Pull request by lpy**
Create the basic framework of GPU profiling validation as well as a gapit
command to run validation. When an Android device is connected, register a
validator if the GPU is supported and the device has basic GPU profiling
capabilities discovered.

BUG: b/138717619
Test: bazel run gapit validate

### CODE REVIEW

**Comment by lpy** [comment]
PTAL!

**Comment by lpy** [comment]
Alright, this time for real. PTAL!

File: gapis/validate/validator/validator.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/google/gapid/gapis/perfetto/service"
+)
+
+type GpuCounter struct {
+	Id    uint32
+	Name  string
+	Check func(column *service.QueryResult_ColumnValues, columnType service.QueryResult_ColumnDesc_Type) bool
+}
+
+// Device represents a connection to an attached device.
```
Comment (hevrard) [inline gapis/validate/validator/validator.go]: This looks like an irrelevant comment, maybe left-over from copy-paste?

File: cmd/gapit/validate.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	}
+	defer client.Close()
+	devices, err := client.GetDevices(ctx)
+	if err != nil {
+		return log.Err(ctx, err, "Failed to get device list.")
+	}
+
+	stdout := os.Stdout
+
+	// TODO: Allow to specify an Android device and ActivityAction.
```
[the remark this answers is not in the document]
Comment (AWoloszyn) [reply cmd/gapit/validate.go]: If you are going to use TODO(bug) use the github issue #

File: cmd/gapit/validate.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	"flag"
+	"fmt"
+	"os"
+
+	"github.com/google/gapid/core/app"
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/core/os/device"
+)
+
+type validateVerb struct{ ValidateFlags }
```
Comment (AWoloszyn) [inline cmd/gapit/validate.go]: I would prefer the verb name to be a bit more descriptive. There are a thousand things we could be validating. 

validate_gpu_profiling or some such.

**Comment by AWoloszyn** [comment]
So, why is this a separate request, rather than putting this information directly in the device info? It seems like a really clunky API to have to

1) Get a device, (which already tells you if it supports perfetto or not)
2) Request if it actually gives you reasonable data when you want to profile.

Why do we have #2. Shouldn't this just be a part of an extended version of #1. The perfetto config already returns what types of profiling information are valid.

Or is this simply a completely separate way to see if the data that comes out of a driver is possibly reasonable at all.

**Comment by pmuetschard** [comment]
> So, why is this a separate request?

Validation will take on the order of 10s of seconds, so we don't want to do it all the time, but only once per device and cache passing results. To simplify things, the UI will do the remembering of which devices/driver versions have been validated. Having it a separate request makes sense to me, but we could make it part of the `get(<device path>)` by adding a boolean to the path.

File: gapis/validate/android/adreno/validate.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+// 2. Missing GPU counter samples
+// 3. Fail to check
+func validateGpuCounters(ctx context.Context, processor *perfetto.Processor) error {
+	for _, counter := range counters {
+		queryResult, err := processor.Query(fmt.Sprintf(counterIdQuery, counter.Name))
+		if err != nil {
+			return log.Errf(ctx, err, "Failed to query with %v", fmt.Sprintf(counterIdQuery, counter.Name))
+		}
+		var counterId int64
+		for _, column := range queryResult.GetColumns() {
```
[the remark this answers is not in the document]
Comment (lpy) [reply gapis/validate/android/adreno/validate.go]: Thanks, added a check above this.

For the first question, yes this is intended. The query we make here is to query the counter id from `counter_definitions` table with the counter name, which should return exactly one long value. Hence we break immediately when we get the `counter_id`

File: gapis/validate/android/validate.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+func NewValidator(dev bind.Device) validator.Validator {
+	gpu := dev.Instance().GetConfiguration().GetHardware().GetGPU()
+	if strings.Contains(gpu.GetName(), "Adreno") {
+		return &adreno.AdrenoValidator{dev.(adb.Device)}
+	}
+
+	// This should never happen since we always avoid registering for
+	// unsupported GPU families.
+	return nil
```
[the remark this answers is not in the document]
Comment (lpy) [reply gapis/validate/android/validate.go]: That won't be necessary. We filter GPU family before calling `NewValidator`, and we also check for `nil` value when we get the validator from manager.

File: gapis/validate/validate.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+	"github.com/google/gapid/core/os/android/adb"
+	"github.com/google/gapid/gapis/perfetto"
+	perfetto_android "github.com/google/gapid/gapis/perfetto/android"
+	"github.com/google/gapid/gapis/service"
+	"github.com/google/gapid/gapis/service/path"
+)
+
+const (
+	activityName   = "android.app.NativeActivity"
+	bufferSizeKb   = uint32(131072)
```
[the remark this answers is not in the document]
Comment (lpy) [reply gapis/validate/validate.go]: kilobytes, I grab this value by dumping a normal perfetto trace config when I start a normal trace

File: gapis/validate/validate.go [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+	mgr := GetManager(ctx)
+	validator, ok := mgr.validators[p.ID.ID()]
+	if !ok || validator == nil {
+		return log.Errf(ctx, nil, "Could not find validator for device %d", p.ID.ID())
+	}
+	d := validator.GetDevice().(adb.Device)
+
+	// Get ActivityAction
+	packages, _ := d.InstalledPackages(ctx)
+	pkg := packages.FindByName(validationPackageName)
```
[the remark this answers is not in the document]
Comment (lpy) [reply gapis/validate/validate.go]: As a baby step, I am hard coding this given that none of our code except `run gapit` will execute this path. The final shape of it would be integrating the cube as part of gapidapk, and hence this package name will eventually be the package name of gapidapk.

**Comment by AWoloszyn** [comment]
It just seems odd to me that the device will claim to support profiling (even GPU profiling), and then we have to make a secondary request to see if it "REALLY" supports GPU profiling.

If this is the case, ANY information about profiling support should be pulled from the device info.

**Comment by lpy** [comment]
chatted offline to understand @AWoloszyn 's concerns.

@AWoloszyn is OK to proceed review. The current decision we agreed on is that I will do a follow up patch to hide the gpu profiling capability when `gapit devices` is run, and return it when `gapit devices -- enable_gpu_profiling` is run.

**Comment by AWoloszyn** [comment]
And to be clear, this should be cleared on the server side.
**and** when we actually ask for it, we should validate at the same time (i..e it should return the correct information)

**Comment by pmuetschard** [comment]
Two comments:

1. Why remove it from the device info? We don't want to validate a device every time. It makes sense to me that the device info contains all the "claimed" profiling capabilities and then the validation returns which of those things we think are valid. The UI will need to know what profiling features the device supports without doing validation in order to cache validation results. Caching here is across GAPID runs, i.e. on disk.

2. I do not think we should introduce yet-another-manager that maps 1:1 to a device. I think it makes sense to put the validation function into the `Tracer` interface. It already contains the `TraceConfiguration` API, so why not a `ValidateTraceConfiguration` API?

**Comment by AWoloszyn** [comment]
@pmuetschard For the device info, it makes me very uneasy to be able to get 2 different results for the question "Does this device support GPU profiling". One of which is "correct" and one of which is "not".

If I cannot rely on the response from device_info, then in my mind, it should not be there. If we have to make 2 requests, then we should just ALWAYS make the second request (or use the cached result) That way the second request can say whether or not it is supported. The "supporst GPU info" in the device-info becomes basically meaningless, other than "if this is true, then you should check something else as well". It seem that it might be easier to always just make the second request.

**Comment by lpy** [comment]
PTAL! I have moved the `Validator` into `tracer/`

File: gapis/trace/tracer/tracer.go [hunk 8]
```diff
@@ -49,6 +51,18 @@ type Process interface {
...
+type GpuCounter struct {
+	Id    uint32
+	Name  string
+	Check func(column *perfetto_service.QueryResult_ColumnValues, columnType perfetto_service.QueryResult_ColumnDesc_Type) bool
+}
+
+type Validator interface {
+	Validate(ctx context.Context, processor *perfetto.Processor) error
+	GetCounterIds() []uint32
+	GetDevice() bind.Device
```
Comment (pmuetschard) [inline gapis/trace/tracer/tracer.go]: The tracer already has this.

File: gapis/trace/tracer/tracer.go [hunk 9]
```diff
@@ -49,6 +51,18 @@ type Process interface {
 	Capture(ctx context.Context, start task.Signal, stop task.Signal, w io.Writer, written *int64) (size int64, err error)
 }
 
+type GpuCounter struct {
+	Id    uint32
+	Name  string
+	Check func(column *perfetto_service.QueryResult_ColumnValues, columnType perfetto_service.QueryResult_ColumnDesc_Type) bool
+}
+
+type Validator interface {
```
Comment (pmuetschard) [inline gapis/trace/tracer/tracer.go]: I think it's better to move this down a level. That is, from the `Tracer`'s perspective, there is just a simple `Validate()` function. This moves all the validation logic to a lower layer. I think validation of a desktop or Stadia device will look drastically different from an Android device.

For now, desktop and Stadia can just have the `Validate()` function return a successful result, while the Android validator can do the whole tracing-processing-query validation loop that is aware of counters, etc.

**Comment by lpy** [comment]
Comments addressed, ptal!

File: gapis/trace/android/adreno/validate.go [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+	return true
+}
+
+func check(ctx context.Context, result *perfetto_service.QueryResult, f checker) bool {
+	columnDescriptor := result.GetColumnDescriptors()[0]
+	column := result.GetColumns()[0]
+	columnType := columnDescriptor.GetType()
+	if columnType != perfetto_service.QueryResult_ColumnDesc_LONG && columnType != perfetto_service.QueryResult_ColumnDesc_DOUBLE {
+		return false
+	}
```
Comment (pmuetschard) [inline gapis/trace/android/adreno/validate.go]: make this a `isNumber` checker, then create a `func add(c1, c2 checker) checker` and use `and(isNumber, isGreaterThanZero)`, etc.

File: gapis/trace/android/trace.go [hunk 11]
```diff
@@ -50,12 +80,114 @@ type androidTracer struct {
...
+			},
+		},
+	}, nil
 }
 
 func (t *androidTracer) GetDevice() bind.Device {
 	return t.b
 }
 
+func (t *androidTracer) Validate(ctx context.Context) error {
```
[the remark this answers is not in the document]
Comment (lpy) [reply gapis/trace/android/trace.go]: I will do it in a follow-up patch. WDYT if I extend and use the Error defined in service.proto?

File: gapis/trace/tracer/tracer.go [hunk 12]
```diff
@@ -49,6 +51,17 @@ type Process interface {
...
+	Validate(ctx context.Context, processor *perfetto.Processor) error
+	GetCounterIds() []uint32
+}
+
```
[the remark this answers is not in the document]
Comment (lpy) [reply gapis/trace/tracer/tracer.go]: there will be `arm` package using `GpuCounter` type so I put it at the same level of `Validator`

[the remark this answers is not in the document]
Comment (lpy) [reply gapis/trace/tracer/tracer.go]: Since `trace/android` depends on trace/android/adreno, `GpuCounter` can't be moved to `trace/android`. Not sure whether there's a different place I can move it, but if we keep it in `tracer`, then I would prefer to keep `Validator` in `tracer` as well.

[the remark this answers is not in the document]
Comment (pmuetschard) [reply gapis/trace/tracer/tracer.go]: I think the packages should be:
- `trace/android/validate` only base deps (contains `Validator` interface and shared code)
- `trace/android/adreno` only base deps and `trace/android/validate`
- `trace/android/arm` only base deps and `trace/android/validate`
- `trace/android` only base deps, `trace/android/validate`, and the impls

If you want to share the counter objects, you can share the entire counter verification, no?

File: gapis/trace/android/trace.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (pmuetschard) [reply gapis/trace/android/trace.go]: Yes extending it is the idea, but for some things you can use the existing ones (e.g. `ErrDataUnavailable` if no validator exists for the given GPU, `ErrInvalidPath` if the device path is bad, etc)

File: gapis/trace/tracer/tracer.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (lpy) [reply gapis/trace/tracer/tracer.go]: thanks! Will do a dedicated `validate/` package.

hmmmmm, not sure what do you mean by counter objects. `mali` and `adreno` expose different counters so no, the counter verification can't be shared. Maybe some checkers can but we can adjust when we see counters from `mali`.

[the remark this answers is not in the document]
Comment (lpy) [reply gapis/trace/tracer/tracer.go]: Mark as resolved.

[the remark this answers is not in the document]
Comment (pmuetschard) [reply gapis/trace/tracer/tracer.go]: What I meant was that you can change the `Validator` interface to return the counter objects instead of just the IDs. You will still create the counters along with their checker functions inside each impl package, but executing the counter queries and all that can be shared.

**Comment by lpy** [comment]
Thanks for the feedback, comments addressed, ptal!

File: gapis/trace/android/trace.go [hunk 13]
```diff
@@ -50,12 +71,134 @@ type androidTracer struct {
...
+	crash.Go(func() {
+		_, err = process.Capture(ctx, startSignal, stopSignal, &buf, &written)
+		doneFunc(ctx)
+	})
+
+	// TODO(lpy): If we start tracing too early, the data has a lot of noise
+	// at the beginning, need to figure out a sweet spot.
+	startFunc(ctx)
+	if !doneSignal.Wait(ctx) {
+		return log.Err(ctx, err, "Fail to wait for done signal from Perfetto.")
```
Comment (pmuetschard) [inline gapis/trace/android/trace.go]: would need a `status.Finish` here, also the `status.Start` below would need the context from before the above `status.Start`.

You can use
```
status.Do(ctx, "thing you're doing", func(ctx context.Context) {
   // put your code here
})
```
It will call the passed function in between start and finish.

I can explain a bit about the quirks of go's `context.Context` on Monday.

File: gapis/trace/android/trace.go [hunk 14]
```diff
@@ -37,11 +44,25 @@ import (
...
 
+const (
+	activityName                            = "com.samples.cube.VKCubeActivity"
+	intentAction                            = "android.intent.action.MAIN"
+	validationPackageName                   = "com.samples.cube"
+	bufferSizeKb                            = uint32(131072)
+	counterPeriodNs                         = uint64(50000000)
+	durationMs                              = 7000
+	gpuCountersDataSourceDescriptorName     = "gpu.counters"
+	gpuRenderStagesDataSourceDescriptorName = "gpu.renderstages"
```
Comment (pmuetschard) [inline gapis/trace/android/trace.go]: go is all about short names. I don't really care, but these are getting a bit long.... :)

File: gapis/trace/android/adreno/validate.go [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+				return false
+			}
+		} else if columnType == perfetto_service.QueryResult_ColumnDesc_DOUBLE {
+			if doubleValues[i] <= 0.0 {
+				return false
+			}
+		}
+	}
+	return true
+}
```
Comment (pmuetschard) [inline gapis/trace/android/adreno/validate.go]: I would
1) make the checker and your three implementations public
2) move all this into the common package.

File: gapis/trace/android/trace.go (hunk 14 above)
Comment (lpy) [reply gapis/trace/android/trace.go]: Please forgive me to ack this one. I am still struggling with abbreviation :/

**Comment by lpy** [comment]
Thanks for the feedback, ptal!

**Comment by lpy** [comment]
gentle ping for review, presubmit should be fixed in the latest patch

File: gapis/trace/android/trace.go [hunk 16]
```diff
@@ -50,12 +71,137 @@ type androidTracer struct {
...
+	defer cleanup.Invoke(ctx)
+
+	// Force to stop the application, ignore any error that happens as it
+	// doesn't affect validation.
+	defer d.ForceStop(ctx, validationPackageName)
+
+	var buf bytes.Buffer
+	var written int64
+	// Start to capture.
+	status.Do(ctx, "Android Device Validation: Tracing", func(ctx context.Context) {
```
Comment (pmuetschard) [inline gapis/trace/android/trace.go]: Don't repeat the name. Just call it `Tracing`. The status package creates a tree, and `Tracing` will be a child of `Android Device Validation`.

ditto below

File: gapis/trace/android/validate/validate.go [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+	"context"
+	"fmt"
+
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/gapis/perfetto"
+	perfetto_service "github.com/google/gapid/gapis/perfetto/service"
+)
+
+const (
+	SampleCounter  = 100
```
Comment (pmuetschard) [inline gapis/trace/android/validate/validate.go]: why public?

**Comment by lpy** [comment]
ptal :)
